src/binder/statement/select_statement.cpp

```diff
@@ -1,37 +1,325 @@
 #include "binder/statement/select_statement.h"
 #include <fmt/format.h>
+#include "binder/expressions/bound_agg_call.h"
+#include "binder/expressions/bound_binary_op.h"
+#include "binder/expressions/bound_column_ref.h"
+#include "binder/expressions/bound_constant.h"
+#include "binder/expressions/bound_star.h"
+#include "binder/expressions/bound_unary_op.h"
+#include "binder/table_ref/bound_base_table_ref.h"
+#include "catalog/catalog.h"
+#include "common/util/string_util.h"
 
 namespace bustub {
 
-SelectStatement::SelectStatement(const Parser &parser, duckdb_libpgquery::PGSelectStmt *pg_stmt)
-    : SQLStatement(StatementType::SELECT_STATEMENT) {
-  bool found = false;
+SelectStatement::SelectStatement(const Catalog &catalog, duckdb_libpgquery::PGSelectStmt *pg_stmt)
+    : SQLStatement(StatementType::SELECT_STATEMENT),
+      table_(make_unique<BoundTableRef>()),
+      where_(make_unique<BoundExpression>()),
+      having_(make_unique<BoundExpression>()),
+      catalog_(catalog) {
+  // bind from clause
+  if (pg_stmt->fromClause != nullptr) {
+    // Extract the table name from the FROM clause.
+    for (auto c = pg_stmt->fromClause->head; c != nullptr; c = lnext(c)) {
+      auto node = reinterpret_cast<duckdb_libpgquery::PGNode *>(c->data.ptr_value);
+      switch (node->type) {
+        case duckdb_libpgquery::T_PGRangeVar: {
+          auto *table_ref = reinterpret_cast<duckdb_libpgquery::PGRangeVar *>(node);
+          table_ = make_unique<BoundBaseTableRef>(table_ref->relname);
+          break;
+        }
+        default:
+          throw Exception(fmt::format("unsupported node type: {}", Binder::NodetypeToString(node->type)));
+      }
+    }
+  } else {
+    table_ = make_unique<BoundTableRef>(TableReferenceType::EMPTY);
+  }
 
-  // Extract the table name from the FROM clause.
-  for (auto c = pg_stmt->fromClause->head; c != nullptr; c = lnext(c)) {
-    auto node = reinterpret_cast<duckdb_libpgquery::PGNode *>(c->data.ptr_value);
-    switch (node->type) {
-      case duckdb_libpgquery::T_PGRangeVar: {
-        if (found) {
-          throw Exception("shouldnt have multiple tables");
+  // bind select list
+  if (pg_stmt->targetList != nullptr) {
+    BindSelectList(pg_stmt->targetList);
+  } else {
+    throw Exception("no select list");
```

Comment:
Generally speaking, I prefer putting the error cases (which tend to be shorter) before the actual logic. This makes it easier to keep track of what's going on IMO.

```
if (blah == nullptr) {
  // error case 1
} else if (blablabla) {
  // error case 2
} else {
  // logic
}
```

IMO, this has the added advantage that if you need to de-indent by one level, it is easy to refactor without breaking anything since the checks are already in the right order.

Reply:
> Generally speaking, I prefer putting the error cases (which tend to be shorter) before the actual logic. This makes it easier to keep track of what's going on IMO.
> 
> ```
> if (blah == nullptr) {
>   // error case 1
> } else if (blablabla) {
>   // error case 2
> } else {
>   // logic
> }
> ```
> 
> IMO, this has the added advantage that if you need to de-indent by one level, it is easy to refactor without breaking anything since the checks are already in the right order.

It turns out that it's due to a clang-tidy lint:

```
/Users/skyzh/Work/bustub/src/binder/statement/select_statement.cpp:48:5: error: do not use 'else' after 'throw' [readability-else-after-return,-warnings-as-errors]
  } else {
    ^~~~~~
```

clang-tidy thinks it's better not to place return / throw branches before actual work. 🤣 I prefer following the clang-tidy rules.

Reply:
That rule is actually to reduce the level of indentation, AFAIK. 

```
if (error checking) {
   ...
}
```

Then you add the code without the else.